### PR TITLE
support autoscaling metrics when deploying models

### PR DESCRIPTION
**Pull request by munagekar**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-aiplatform/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Comment by munagekar** [comment]
@sararob I see that you have added the do-not-merge label to all open PRs. Let me know if there's anything specific that you want addressed in this PR.

My team needs this feature to be added to the sdk to use vertex AI.

**Comment by munagekar** [comment]
@sasha-gitg Could you take a look ?

File: google/cloud/aiplatform/models.py [hunk 1]
```diff
@@ -980,6 +1011,12 @@ def _deploy_call(
                 "Both `accelerator_type` and `accelerator_count` should be specified or None."
             )
 
+        if not accelerator_type or not accelerator_count and autoscaling_target_accelerator_duty_cycle:
```
Comment (sararob) [inline google/cloud/aiplatform/models.py:1014]: I don't think this check is working as expected (I tried deploying a model with only `autoscaling_target_cpu_utilization` set and it raised this error). You can change it to something like:

```
if autoscaling_target_accelerator_duty_cycle is not None and (not accelerator_type or not accelerator_count):
```

File: google/cloud/aiplatform/models.py [hunk 2]
```diff
@@ -840,6 +853,13 @@ def _deploy(
                 be immediately returned and synced when the Future has completed.
             deploy_request_timeout (float):
                 Optional. The timeout for the deploy request in seconds.
+            autoscaling_target_cpu_utilization (int):
+                Target CPU Utilization to use for Autoscaling Replicas.
```
Comment (sararob) [inline google/cloud/aiplatform/models.py:857]: Could you add "Optional" to the beginning of this docstring? Same for `autoscaling_target_accelerator_duty_cycle`.

**Review by sararob — CHANGES_REQUESTED** [review]
Thanks for this contribution! Could you add unit tests in [`tests/unit/aiplatform/test_endpoints.py`](https://github.com/googleapis/python-aiplatform/blob/main/tests/unit/aiplatform/test_endpoints.py) for these additions? You can add the following tests:
* test_deploy_with_autoscaling_target_cpu_utilization
* test_deploy_with_autoscaling_target_accelerator_duty_cycle
* test_deploy_with_autoscaling_target_accelerator_duty_cycle_and_no_accelerator_type_or_count_raises (to ensure this raises a `ValueError`)

Let me know if you need any help adding these.

File: google/cloud/aiplatform/models.py (hunk 2 above)
Comment (munagekar) [reply google/cloud/aiplatform/models.py:857]: Fixed in https://github.com/googleapis/python-aiplatform/pull/1197/commits/7a601c7266c9e871c5607b6375bfbbfc28be992b

File: google/cloud/aiplatform/models.py (hunk 1 above)
Comment (munagekar) [reply google/cloud/aiplatform/models.py]: Fixed in https://github.com/googleapis/python-aiplatform/pull/1197/commits/9501eb0904ee10a5c7a35425cb21c732659c5a3c

**Comment by munagekar** [comment]
> Could you add unit tests in [tests/unit/aiplatform/test_endpoints.py](https://github.com/googleapis/python-aiplatform/blob/main/tests/unit/aiplatform/test_endpoints.py) for these additions?

Thankyou for the review. I have added in tests. PTAL

**Comment by munagekar** [comment]
> Could you add unit tests in [tests/unit/aiplatform/test_endpoints.py](https://github.com/googleapis/python-aiplatform/blob/main/tests/unit/aiplatform/test_endpoints.py) for these additions?

@sararob Thankyou for the review. I have added in tests. PTAL

File: tests/unit/aiplatform/test_endpoints.py [hunk 3]
```diff
@@ -1057,6 +1057,131 @@ def test_deploy_with_dedicated_resources(self, deploy_model_mock, sync):
...
+            )
+            test_endpoint.deploy(
+                model=test_model,
+                machine_type=_TEST_MACHINE_TYPE,
+                service_account=_TEST_SERVICE_ACCOUNT,
+                sync=sync,
+                deploy_request_timeout=None,
+                autoscaling_target_accelerator_duty_cycle=70
+            )
+
```
Comment (sararob) [inline tests/unit/aiplatform/test_endpoints.py:1184]: Could you add the following after the `deploy` call so the test passes with `sync=False`?

```
if not sync:
    test_endpoint.wait()
```

File: tests/unit/aiplatform/test_endpoints.py [hunk 4]
```diff
@@ -1057,6 +1057,131 @@ def test_deploy_with_dedicated_resources(self, deploy_model_mock, sync):
...
+    ):
+        with pytest.raises(ValueError):
+            test_endpoint = models.Endpoint(_TEST_ENDPOINT_NAME)
+            test_model = models.Model(_TEST_ID)
+            test_model._gca_resource.supported_deployment_resources_types.append(
+                aiplatform.gapic.Model.DeploymentResourcesType.DEDICATED_RESOURCES
+            )
+            test_endpoint.deploy(
+                model=test_model,
+                machine_type=_TEST_MACHINE_TYPE,
```
Comment (sararob) [inline tests/unit/aiplatform/test_endpoints.py:1178]: You can remove the `machine_type`, `service_account`, and `deploy_request_timeout` parameters from this call.

File: tests/unit/aiplatform/test_endpoints.py [hunk 5]
```diff
@@ -1057,6 +1057,131 @@ def test_deploy_with_dedicated_resources(self, deploy_model_mock, sync):
...
+
+        if not sync:
+            test_endpoint.wait()
+
+        expected_machine_spec = gca_machine_resources.MachineSpec(
+            machine_type=_TEST_MACHINE_TYPE,
+        )
+
+        expected_autoscaling_metric_spec = gca_machine_resources.AutoscalingMetricSpec(
+            metric_name="aiplatform.googleapis.com/prediction/online/cpu/utilization",
```
Comment (sararob) [inline tests/unit/aiplatform/test_endpoints.py:1085]: Could you save this `metric_name` string to variable at the top of the file, something like `_TEST_METRIC_NAME_CPU_UTILIZATION`?

**Review by sararob — CHANGES_REQUESTED** [review]
Thanks for adding tests. Left a few comments, once these are addressed this should be ready to merge after presubmit tests pass.

**Comment by munagekar** [comment]
Thank you for reviewing the changes. I don't have enough bandwidth this week. I will send in patches early next week.

File: tests/unit/aiplatform/test_endpoints.py (hunk 3 above)
Comment (munagekar) [reply tests/unit/aiplatform/test_endpoints.py:1184]: No of the other raises tests which raise errors wait for the operation to complete. 

I think a Value Error will take place and `test_endpoint` will be undefined.

File: tests/unit/aiplatform/test_endpoints.py (hunk 5 above)
Comment (munagekar) [reply tests/unit/aiplatform/test_endpoints.py]: Fixed in https://github.com/googleapis/python-aiplatform/pull/1197/commits/e14d1ae44bf5ff51b91f4011f25f38aeac84a4db

File: tests/unit/aiplatform/test_endpoints.py (hunk 4 above)
Comment (munagekar) [reply tests/unit/aiplatform/test_endpoints.py]: Fixed in https://github.com/googleapis/python-aiplatform/pull/1197/commits/fbceb92d98a05cc6e992bcaed723a5228808bef8

File: tests/unit/aiplatform/test_endpoints.py [hunk 6]
```diff
@@ -1057,6 +1057,131 @@ def test_deploy_with_dedicated_resources(self, deploy_model_mock, sync):
...
+            test_endpoint.wait()
+
+        expected_machine_spec = gca_machine_resources.MachineSpec(
+            machine_type=_TEST_MACHINE_TYPE,
+            accelerator_type=_TEST_ACCELERATOR_TYPE,
+            accelerator_count=_TEST_ACCELERATOR_COUNT,
+        )
+
+        expected_autoscaling_metric_spec = gca_machine_resources.AutoscalingMetricSpec(
+            metric_name="aiplatform.googleapis.com/prediction/online/accelerator/duty_cycle",
```
[the remark this answers is not in the document]
Comment (munagekar) [reply tests/unit/aiplatform/test_endpoints.py]: Fixed in https://github.com/googleapis/python-aiplatform/commit/e14d1ae44bf5ff51b91f4011f25f38aeac84a4db

**Comment by munagekar** [comment]
@sararob I have added review comments. PTAL.

File: tests/unit/aiplatform/test_endpoints.py (hunk 3 above)
Comment (sararob) [reply tests/unit/aiplatform/test_endpoints.py:1184]: Yes, this test needs the wait call. The ones that don't need it are testing invalid parameter values (negative numbers and > 100 for `traffic_percentage` so the LRO will never start in those cases.

**Review by sararob — CHANGES_REQUESTED** [review]
After adding the `wait()` call in `test_deploy_with_autoscaling_target_accelerator_duty_cycle_and_no_accelerator_type_or_count_raises`, this should be good to go.

File: tests/unit/aiplatform/test_endpoints.py (hunk 3 above)
Comment (munagekar) [reply tests/unit/aiplatform/test_endpoints.py:1184]: Fixed in https://github.com/googleapis/python-aiplatform/pull/1197/commits/8a28d7ae459f36bb72eba27cf14f6b3ef0f9abf5

**Comment by munagekar** [comment]
@sararob Thank you for review and clarifications. Updated the PR, PTAL.

**Review by sararob — CHANGES_REQUESTED** [review]
The unit tests passed, you just need to run the linter. You can do this in the root of the repo with:

```
pip3 install nox && nox -s lint
```

And then push a commit with the changes. Or if you give me push access to your branch I can run it and merge the PR. Thanks!

**Comment by munagekar** [comment]
@sararob Thank you for the review.I have fixed the linting error, lint check seems to pass locally.

- I have already allowed edit by maintainers of googleapis/python-aiplatform for this PR.
- Additionally I have sent a collaborator request to you for `munagekar/python-aiplatform`.

Feel free to take over the PR if there are any other issues.

**Review by sararob — APPROVED** [review]
